### PR TITLE
dynamic_forward_proxy: adding dns_resolvers to dns_cache used by the proxy filter

### PR DESCRIPTION
**Pull request by ntgsx92**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

Commit Message:
Adding `dns_resolvers` option to the dns cache used by the dynamic forward proxy filter. This enables users to switch default dns resolvers from `/etc/resolv.conf` to custom dns resolvers in the proxy filter. By default, this option will be empty which means dns cache will continue to resolve names from servers listed in `/etc/resolv.conf`.

Additional Description:
None

Risk Level:
Low

Testing:
Unit test added

Docs Changes:
Yes

Release Notes:
Added

Platform Specific Features:
N/A

[Optional Runtime guard:]
[Optional Fixes #Issue]

https://github.com/envoyproxy/envoy/issues/15657
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto
Comment (adisuissa) [inline api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:111]: nit: Please add what will be used if not specified (i.e., using the default resolver).

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this.
One small comment about the API.

**Comment by malza321** [comment]
> 
> 
> For an explanation of how to fill out the fields, please see the relevant section
> 
> in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
> 
> 
> 
> Commit Message:
> 
> Adding `dns_resolvers` option to the dns cache used by the dynamic forward proxy filter. This enables users to switch default dns resolvers from `/etc/resolv.conf` to custom dns resolvers in the proxy filter. By default, this option will be empty which means dns cache will continue to resolve names from servers listed in `/etc/resolv.conf`.
> 
> 
> 
> Additional Description:
> 
> None
> 
> 
> 
> Risk Level:
> 
> Low
> 
> 
> 
> Testing:
> 
> Unit test added
> 
> 
> 
> Docs Changes:
> 
> Yes
> 
> 
> 
> Release Notes:
> 
> Added
> 
> 
> 
> Platform Specific Features:
> 
> N/A
> 
> 
> 
> [Optional Runtime guard:]
> 
> [Optional Fixes #Issue]
> 
> 
> 
> https://github.com/envoyproxy/envoy/issues/15657
> 
> [Optional Deprecated:]
> 
> [Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]
> 
> 

You are my hero!

File: api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto
Comment (ntgsx92) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: Thanks. Fixed the description.

**Review by adisuissa — COMMENTED** [review]
API lgtm.
@htuch can you please take a quick look at the API changes?

File: api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto
Comment (htuch) [inline api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: How does this relate to https://github.com/envoyproxy/envoy/pull/16237? Should this be a `repeated DnsResolver` and `DnsResolver` has both address and `DnsLookupOptions`?

Comment (ntgsx92) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: Thanks for letting me know about this change.

Looks like we're going to have `resolvers`, `use_tcp_for_dns_lookups` and `no_defalt_search_domain` as the arguments for `DnsResolverImpl` after #16237 is merged.
Seems like a good time to encapsulate those fields into a single proto message like `DnsResolver`.

@suniltheta How do you feel about the above changes?

Comment (suniltheta) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: Thanks for letting me know about this PR. Just recently committed all of the changes for the PR #16237. A new protobuf message DnsResolverOptions in introduced which only include bool fields use_tcp_for_dns_lookups & no_default_search_domain.

Notice the name change from DnsLookupOptions to DnsResolverOptions.

Having said that if we move resolvers inside DnsResolverOptions, then we might have to take care of deprecating dns_resolvers present in config/cluster/v3/cluster.proto.

I lean towards keeping the resolvers (repeated Address type) out of dedicated DnsResolverOptions. Only because if we combine we are looking at a slightly bloated PR and deprecation of a field. But I am also flexible in combining them, if doing so is the best option in the long run.

Comment (htuch) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: Yeah, I would separate `DnsResolverOptions` and reference it from a `DnsResolver` message, which can deal with addressing and other concerns not related to resolution options.

Comment (suniltheta) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: So is this the format of the protobuf message that is being proposed?
```
└── DnsResolver
    └── <repeated> Address
    └── DnsResolverOptions
        ├── <bool> use_tcp_for_dns_lookups
        └── <bool> no_default_search_domain
        └── /<any options added in future>/
```

After the merge of #16237 we do
`dispatcher().createDnsResolver(resolvers, dns_resolver_options);`

Instead we would be doing
`dispatcher().createDnsResolver(dns_resolvers);`
So when we call the function `createDnsResolver ` we would be only passing the object of type `DnsResolver`. Here `dns_resolvers` will contain both `resolvers` and `dns_resolver_options`.

Comment (suniltheta) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: Or is this the format
**FORMAT B**
```
├── DnsResolver
│   ├── <repeated Address> dns_resolvers
│   └── /<other concerns>/
└── <DnsResolverOptions> dns_resolver_options
        ├── <bool> use_tcp_for_dns_lookups
        ├── <bool> no_default_search_domain
        └── /<any options added in future>/
```
Finally the function call would be 
`dispatcher().createDnsResolver(dns_resolver, dns_resolver_options);`

Comment (ntgsx92) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: Sounds good. How do we want to proceed from here? I think we can either,

1) Wait for #16237 to be merged which adds `DnsResolverOptions` message and I can update the PR to create a new message `DnsResolver` and reference `DnsResolverOptions` from there.

2) Merge this PR as it is and have another PR that adds `DnsResolver ` message.

Comment (suniltheta) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: I think we may not be able to merge #16237 or this PR without resolving the merge conflicts in the one which gets merged afterwards.
I am ok with either options, it is either one of us who would be taking care of merge conflicts. Let me know your preference and I can stick with that.

I am inclined to making "a new message `DnsResolver`" in a separate PR (different from this current PR) so that it will easy for review and nothing else.

Comment (ntgsx92) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: I think `FORMAT A:` is a better layout as it directly maps to a list of arguments one would need to construct a `DnsResolver`.

Comment (ntgsx92) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: @htuch Let us know if you're okay with the above plan. Thanks!

Comment (htuch) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: Do we have `repeated DnsResolver` at top-level? I think each `DnsResolver` should have an `Address` and also a `DnsResolverOptions` perhaps.

Comment (ntgsx92) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: So instead of
```
└── DnsResolver
    ├── <repeated Address> dns_resolvers
    └── <DnsResolverOptions> dns_resolver_options
        ├── <bool> use_tcp_for_dns_lookups
        ├── <bool> no_default_search_domain
        └── /<any options added in future>/
```
We can have
```
└── repeated DnsResolver
    ├── <Address> dns_resolver
    └── <DnsResolverOptions> dns_resolver_options
        ├── <bool> use_tcp_for_dns_lookups
        ├── <bool> no_default_search_domain
        └── /<any options added in future>/
```
?
SGTM

Comment (suniltheta) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: When we initialize the [DnsResolver](https://github.com/envoyproxy/envoy/blob/main/source/common/network/dns_impl.cc#L30) we have the ability to specify the DNS resolver lookup options once and send a list of external DNS servers (ip addresses) the resolver can query. I believe in the core implementation of c-ares resolver we call on this function [ares_set_servers_ports_csv](https://c-ares.haxx.se/ares_set_servers_ports_csv.html) on an already initialized channel to supply a string containing list of IP address. 

I do not see the benefit of specifying distinct resolver options for each external DNS server address. Please correct me if my assumptions are wrong.

Also @htuch is your suggestion on `repeated DnsResolver` is related to having the ability to specify the multiple DNS resolver libraries to be used? Like envoy currently uses c-ares as the core Dns Resolver and tomorrow there can be extensions added for a new Dns resolver (Eg: udns, adns etc). In that case should we add a field which says what DNS resolver library to be initialized (c-ares or other DNS resolver extensions) ?

> Sorry about dragging the conversation on a PR comment. Should we create an github issue (design proposal) and discuss there?

Comment (htuch) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: Yeah, fair point. Let's go with https://github.com/envoyproxy/envoy/pull/16294#discussion_r627609104 then.

Comment (ntgsx92) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: @htuch What's the next step? @suniltheta and I discussed merging this PR as it is and make another PR to add the new `DnsResolver` message.

Comment (htuch) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: If we're planning on immediately restructuring in the next PR that isn't so great. Can we at least introduce the `DnsResolver` message and place the list of addresses in there?

Comment (suniltheta) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: I am on board with this idea as well. We can place the `DnsResolver` protobuf message in `api/envoy/config/core/v3/base.proto` or any other place which makes more sense in this PR. Move `repeated Address dns_resolvers` inside `DnsResolver` like mentioned in previous comment. Finally as part of #16237 PR we can move the `DnsResolverOptions` inside the already merged `DnsResolver` message.

Comment (htuch) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:113]: +1

File: api/envoy/config/core/v3/base.proto
Comment (suniltheta) [inline api/envoy/config/core/v3/base.proto:454]: nit: Should we just name it just `resolvers` instead of `dns_resolvers` ?

Comment (ntgsx92) [reply api/envoy/config/core/v3/base.proto]: updated

**Comment by malza321** [comment]
hi!

Can I see how to use custom resolvers in advance? 
envoy.yaml will be great help to let me understand!

Thanks!!

File: api/envoy/config/core/v3/base.proto
Comment (htuch) [inline api/envoy/config/core/v3/base.proto:448]: Nit: suggest moving this to its own file `resolver.proto` in `api/envoy/config/v3`.

**Review by htuch — COMMENTED** [review]
API LGTM modulo nit.

File: api/envoy/config/core/v3/base.proto
Comment (suniltheta) [reply api/envoy/config/core/v3/base.proto:448]: While we are at this, what does others think about renaming `DnsResolver` to `DnsResolverConfig` ? So that we don't get confused between this message and a class of the same name [DnsResolver](https://github.com/envoyproxy/envoy/blob/main/include/envoy/network/dns.h#L44). Apologies for not bringing this up earlier.

Comment (htuch) [reply api/envoy/config/core/v3/base.proto:448]: Sure, although I think that's a good idea just because the naming scheme is clearer; avoiding C++ level type conflicts isn't that strong a motivator, we have namespaces etc to deal with that.

**Comment by ntgsx92** [comment]
@htuch the presubmit script kept [failing](headers_with_underscores_action) for some reasons. Do you why it's failing?

**Comment by suniltheta** [comment]
Can you confirm if [protodoc-title](https://github.com/envoyproxy/envoy/pull/16294/commits/f9a8326f0bb91ac46f1a6e6c43c60aed765cce3d) change got applied to api/envoy/config/core/v4alpha/resolver.proto file as well? 

ref: Run `./tools/proto_format/proto_format.sh fix` from [STYLE.md](https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md)

File: api/envoy/config/core/v3/base.proto
Comment (malza321) [reply api/envoy/config/core/v3/base.proto]: is it also possible to foward ip directly without dns query?

I did curl -x http://envoyip:port -kLv http://targetip:port but envoy keep returning "No healthy upstream"

**Review by htuch — COMMENTED** [review]
/lgtm api

File: api/envoy/config/core/v3/base.proto
Comment (htuch) [reply api/envoy/config/core/v3/base.proto]: So, definitely open to (maybe it's clearer) to call this `DnsResolverConfig` or `DnsResolutionConfig`.

**Comment by htuch** [comment]
@suniltheta looks like it's passing format CI check.

File: api/envoy/config/core/v3/base.proto
Comment (suniltheta) [reply api/envoy/config/core/v3/base.proto]: let's say, we can take that up renaming `DnsResolver` --> `DnsResolverConfig` or `DnsResolutionConfig` in the next PR #16237 ? Since all the CI got passed here, this can go?

**Comment by suniltheta** [comment]
> @suniltheta looks like it's passing format CI check.

Yes, it is passing all the checks now. Thanks @ntgsx92 for taking care of it.

**Comment by ntgsx92** [comment]
@adisuissa I think we should be good with merging this change. Do you have any other concerns? Thanks!

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc
Comment (adisuissa) [inline source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:60]: nit: can pass directly to the parameter.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

Added a small nit to the code.

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc
Comment (suniltheta) [reply source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:60]: Hi @adisuissa, since the PR #16237 will follow this PR(16294) shortly, this line will be replaced by using `dns_resolver_options`. Can this PR get merged as it is?

**Review by adisuissa — DISMISSED** [review]
I think it looks good.

/assign-from @envoyproxy/first-pass-reviewers

**Comment by ntgsx92** [comment]
Looks like this test case is [broken](https://dev.azure.com/cncf/envoy/_build/results?buildId=76369&view=logs&j=1cf12c52-e2c4-5324-3db9-61ecde950a49&t=aec3565d-6fc0-512c-7d36-23d30fe2a5b2&l=4192). Not sure why this PR will cause it to break.

**Comment by ntgsx92** [comment]
@junr03 Hey, mind taking a look at PR? Thanks! 😄

File: api/envoy/config/core/v3/resolver.proto
Comment (mattklein123) [inline api/envoy/config/core/v3/resolver.proto:16]: nit: end sentences with periods. Same below and in other API/doc files.

File: api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto
Comment (mattklein123) [inline api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:108]:
```suggestion
  // Otherwise, the default system resolvers (e.g., /etc/resolv.conf) will be used.
```

File: api/envoy/config/core/v3/resolver.proto
Comment (mattklein123) [inline api/envoy/config/core/v3/resolver.proto:22]: I would add a validation constraint that this is greater > 1 in size. Then you can drop the code checks below.

File: test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc
Comment (mattklein123) [inline test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc:97]: `expected_resolvers`

File: test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc
Comment (mattklein123) [inline test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc:108]: Why are there all these deletions?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small comments. Please also merge main.

/wait

File: test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc
Comment (ntgsx92) [reply test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc:108]: fixed. Thanks

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**
